src/visitors/hoist.js

```diff
@@ -355,16 +327,6 @@ function addExport(asset, path, local, exported) {
   asset.cacheData.exports[getName(asset, 'export', exported.name)] =
     exported.name;
 
-  if (path.scope.hasGlobal('module') || path.scope.hasGlobal('exports')) {
-    path.insertAfter(
-      EXPORT_ASSIGN_TEMPLATE({
-        EXPORTS: getExportsIdentifier(asset),
-        NAME: t.identifier(local.name),
-        LOCAL: getIdentifier(asset, 'export', exported.name)
-      })
-    );
-  }
-
```

Comment:
I think removing this causes an issue where if you reference commonjs exports before an es6 export it will insert the es6 export in the wrong place (before it is initialized, so it will be undefined). I've added a failing test case for that.

```javascript
(function () {
  var $4$exports = {
    b: $4$export$b
  };


  /* ASSET: 4 - b.js */
  $4$exports.a = 2;
  var $4$export$b = 3;

  /* ASSET: 2 - a.js */
  var $2$var$_require = $4$exports,
      $2$var$a = $2$var$_require.a,
      $2$var$b = $2$var$_require.b;

  output = $2$var$a + $2$var$b;
})();
```

As you can see, `$4$export$b` is referenced before it is initialized since the commonjs export comes first.

Reply:
The easiest way to fix this would be to use getters for all exports (and ignore if the binding is constant). 

```js
var $4$exports = {
  get b() { return $4$export$b }
};
```

I have no idea if it has an impact on the final size though.

Reply:
Yeah that would work... All those getters would probably hurt minifiers being able to statically resolve the values of object properties and potentially get rid of the object all together though right?

Why not just do what we did before which was insert the empty object at the top if needed, and assign to it as you go?

Reply:
Not necessarily:
```
$ echo 'var o = {a: 1, b: 2}; console.log(o.a, o.b, o.a + o.b);' | bin/uglifyjs --toplevel -mc passes=2
console.log(1,2,3);
```

Reply:
Curiously without the IIFE it will be simplified:
```
$ cat t2.js 

    var willNeverBeTrimmed = {};
    willNeverBeTrimmed.one = 'it does not';
    willNeverBeTrimmed.two = 'get trimmed';
    console.log(willNeverBeTrimmed.one);
    function nothingToDoWithTheRest() {
        this_fn_should(make_uglify(bailout));
    }

$ cat t2.js | bin/uglifyjs --toplevel -mc passes=3
console.log("it does not");
```
However, with a surrounding IIFE it is not simplified. That's probably a minor scope issue that could be reported to the uglify project.

It's worth noting that Rollup uses objects for its default exports and handles some real world cases pretty well:

https://github.com/rollup/rollup/issues/501#issuecomment-355654923

The only real point of my comments is that there are more optimization opportunities without getters.

Reply:
@fathyb I'd recommend using the following uglify options for advanced optimization:
```js
compress: {
    passes: 3,
    pure_getters: true, // avoid if code has getters with side effects
    toplevel: true, // not needed if all code wrapped in function scope
    unsafe: true, // optional: actually safe for sane JS not modifying builtins
}
```
Keep in mind that each additional pass increases the minification time. Something to be aware of when selecting defaults.